### PR TITLE
chore(pre-commit): replace several linter by ruff

### PR DESCRIPTION
**Pull request by finswimmer**
This PR replaces several linter by ruff. `ruff` is a very fast linter and autoformatter.

Some unused variables were found by ruff. So I removed them.

I would suggest to register the repo on https://pre-commit.ci now, to run `pre-commit` in every PR.

### CODE REVIEW

**Comment by oz123** [comment]
@finswimmer can you please resolve the conflicts?

**Comment by finswimmer** [comment]
> @finswimmer can you please resolve the conflicts?

This was easier said than done :sweat_smile: Pydantic doesn't support syntax like `str | None`. It took a while to find this out. See https://github.com/pydantic/pydantic/issues/2597

Due to  this, we are now in the situation where we have places that uses the syntax and places where we don't. I tend forward to the idea to revert the future annotations syntax completely for the sake of consistency. What do you think?

**Comment by oz123** [comment]
Well, I only looked at it briefly. But if the only solution to upgrade to Python3.10 we can't merge this PR.
We should keep using the `Optional` syntax instead the set `|` syntax.
Why did you remove them btw? Is it just because ruff suggested it?

**Comment by finswimmer** [comment]
> Why did you remove them btw? Is it just because ruff suggested it?

This is done by `pyugrade` (either the original in the current pre-commit config or the reimplemented one by ruff) due to the `from __future__ import annotations`.

`ruff`'s behavior is better here. In conjunction with the `runtime-evaluated-base-classes` setting of `tool.ruff.flake8-type-checking` it doesn't change the syntax for the pydantic models. One could also disable the [UP038](https://beta.ruff.rs/docs/rules/#pyupgrade-up) check at all. But this will lead to inconsistent usage in the code base over all.

TL;DR: The ruff config in this PR works now and in the future. (And works better than the original `pyupgrade`)

**Comment by oz123** [comment]
OK. So I understand why you remove Optional.

Well, the new | syntax is only used in the function signatrues, and the Optional syntax is in type declarations.
Looks OK to me.

Besides:
> A foolish consistency is the hobgoblin of little minds ...
By R.W Emerson.

We start now, and eventually will remove Optional. I am fine with this.

**Review by oz123 — APPROVED** [review]

**Comment by finswimmer** [comment]
@oz123 : Gently reminder ;) :

> I would suggest to register the repo on https://pre-commit.ci/ now, to run pre-commit in every PR.

**Comment by oz123** [comment]
I don't understand why this is needed. Pre commit runs locally right?
Why another dependency? we have GH actions already.

**Comment by finswimmer** [comment]
> I don't understand why this is needed. Pre commit runs locally right?

`pre-commit` only runs locally if:

* the contributor runs it manually before a commit,
* the contributor has run `pre-commit install` once, so that it runs on every commit automatically,
* the contributor does not by-pass the pre-commit hooks by adding `--no-verify` during commit.

You see there are a lot of chances, that the pre-commit hook do not run before a commit. This is why it should be checked within the CI pipeline. It's absolutely fine to do this via a GH actions.  https://pre-commit.ci/ has the benefit, that there is no need to configure anything and you will receive PR for updates of the hooks.

**Comment by oz123** [comment]
Well, if they chose not to run the pre-commit hook the CI fails. So, there is no risk for us. Just the annoyance of telling them to run it and fix the issue.
As the project is low key - i.e. not many PR\Spam PR I guess we can go with it.
If this starts annoying we can add it.
I just don't want to waster CPU cycles for running both pre-commit and gh actions. It has a small negative environmental effect.

**Comment by matteius** [comment]
@oz123 to be fair ruff replaces a more costly linter to run, so it should be fewer CPU cycles overall.

**Comment by oz123** [comment]
And we already run ruff. I just don't see the point in running GH actions and another CI pipline.
